sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/export/CollectionHandle.java

```diff
@@ -0,0 +1,156 @@
+ * A handle for a collection-pipeline of metrics.
+ *
+ * <p>This class provides an efficient means of leasing and tracking exporters.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public final class CollectionHandle {
+  /** The index of this handle. */
+  private final int index;
+
+  private CollectionHandle(int index) {
+    this.index = index;
+  }
+
+  /** Construct a new (efficient) mutable set for tracking collection handles. */
+  public static Set<CollectionHandle> mutableSet() {
+    return new CollectionHandleSet();
+  }
+
+  /**
+   * Construct a new (mutable) set consistenting of the passed in collection handles.
+   *
+   * <p>Used by tests.
+   */
+  static Set<CollectionHandle> of(CollectionHandle... handles) {
+    Set<CollectionHandle> result = mutableSet();
+    for (CollectionHandle handle : handles) {
+      result.add(handle);
+    }
+    return result;
+  }
+
+  /**
+   * Construct a new supplier of collection handles.
+   *
+   * <p>Handles returned by this supplier should not be used with unique handles produced by any
+   * other supplier.
+   */
+  public static Supplier<CollectionHandle> createSupplier() {
+    return new Supplier<CollectionHandle>() {
+      private final AtomicInteger nextIndex = new AtomicInteger(1);
+
+      @Override
+      public CollectionHandle get() {
+        return new CollectionHandle(nextIndex.getAndIncrement());
+      }
+    };
+  }
+
+  @Override
+  public int hashCode() {
+    return index;
+  }
+
+  @Override
+  public boolean equals(@Nullable Object other) {
+    if (this == other) {
+      return true;
+    }
+    if (other == null) {
+      return false;
+    }
+    if (!(other instanceof CollectionHandle)) {
+      return false;
+    }
+    return index == ((CollectionHandle) other).index;
+  }
+
+  @Override
+  public String toString() {
+    return "CollectionHandle(" + index + ")";
+  }
+
+  /** An optimised bitset version of {@code Set<CollectionHandle>}. */
+  private static class CollectionHandleSet extends AbstractSet<CollectionHandle> {
+    private final BitSet storage = new BitSet();
+
+    @Override
+    public Iterator<CollectionHandle> iterator() {
+      return new MyIterator();
+    }
+
+    @Override
+    public boolean add(CollectionHandle handle) {
+      if (storage.get(handle.index)) {
+        return false;
+      }
+      storage.set(handle.index);
+      return true;
+    }
+
+    @Override
+    public boolean contains(Object handle) {
+      if (handle instanceof CollectionHandle) {
+        return storage.get(((CollectionHandle) handle).index);
+      }
+      return false;
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> other) {
+      if (other instanceof CollectionHandleSet) {
+        BitSet result = new BitSet();
+        BitSet otherStorage = ((CollectionHandleSet) other).storage;
+        result.or(storage);
+        result.and(otherStorage);
```

Comment:
```suggestion
        BitSet result = storage.clone().and(((CollectionHandleSet) other).storage);
```

`BitSet.clone()` even has javadoc!

Reply:
bad habits of never trusting clone...

Also, clone returns `Object` so you have to cast here.

Reply:
Fixed.